### PR TITLE
Added DPD signatures and a gitignore.

### PR DESCRIPTION
**Pull request by keithjjones**
I added a DPD signature and a gitignore.  I looked at the pcap in this testing directory and wrote the signatures based on that.  I then tested it on a live network and things looked good.

### CODE REVIEW

**Comment by keithjjones** [comment]
A note:  I ran into a few false positives with SNMP and NFS on a network I monitor with this logic.  I'm not entirely sure how to fix that.  Maybe some &requires are needed in the spicy code to enforce ldap constructs?  With ASN.1, the dpd sigs are not trivial and using ^0x30 for both sides of the connection is not enough alone to filter out false positives.

File: analyzer/__load__.zeek [hunk 1]
```diff
@@ -1 +1,2 @@
+@load-sigs ./dpd.sig
```
Comment (bbannier) [inline analyzer/__load__.zeek:1]: We need to add `dpd.sig` to the `SCRIPTS` in `analyzer/CMakeLists.txt`.

File: analyzer/dpd.sig [hunk 2]
```diff
@@ -0,0 +1,23 @@
+signature dpd_ldap_client_udp {
```
Comment (bbannier) [inline analyzer/dpd.sig:1]: We should really add a test for DPD. Like you mentioned offline, we could rewrite the PCAP we already have in tree to use a different port and check whether we get similar results.

**Review by bbannier — COMMENTED** [review]
Looks in general good to me, but for minor stuff (I fixed these issues in #3 which you could just cherry-pick into your branch).

I am not sure what to do about false positives you mentioned. Does this lead to us misclassifying traffic which would previously have been matched by another analyzer, or do we now just incorrectly pick up previously unattributed traffic?

File: analyzer/dpd.sig (hunk 2 above)
Comment (keithjjones) [reply analyzer/dpd.sig:1]: Added

File: analyzer/__load__.zeek (hunk 1 above)
Comment (keithjjones) [reply analyzer/__load__.zeek:1]: Added

**Comment by keithjjones** [comment]
I added a test.  It should be good to go now.

**Review by bbannier — APPROVED** [review]

**Comment by bbannier** [comment]
Thanks for the adjustments. I'll merge after CI passes.

**Comment by bbannier** [comment]
I hear that this still produces false positives. I'll wait with merging this to allow some more time for feedback.

**Comment by keithjjones** [comment]
I added a port filter because this will false positive on SNMP, probably because they both use ASN.1.  This is an easy way to stop FPs, but someone should spend time on ldap.spicy to reject anything that isn't LDAP (like SNMP) to do this properly.  Detecting LDAP on all ports but one is much better than just the LDAP ports it currently does, IMHO.

**Comment by mmguero** [comment]
I'll test this out this afternoon as well. I imagine [other protocols](https://en.wikipedia.org/wiki/ASN.1#Applications) that use ASN.1 would also have problems, right? E.g., kerberos, etc.?

**Comment by keithjjones** [comment]
> I'll test this out this afternoon as well. I imagine [other protocols](https://en.wikipedia.org/wiki/ASN.1#Applications) that use ASN.1 would also have problems, right? E.g., kerberos, etc.?

I would assume so.  I'm still new to ASN.1.

**Comment by mmguero** [comment]
Running it through a bunch of PCAPs right now, this kerberos traffic caused a false positive
[krb5.pcap.gz](https://github.com/zeek/spicy-ldap/files/7721700/krb5.pcap.gz)
.
```
> cat krb5.pcap/dpd.log 
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#path	dpd
#open	2021-12-15-17-53-34
#fields	ts	uid	id.orig_h	id.orig_p	id.resp_h	id.resp_p	proto	analyzer	failure_reason
#types	time	string	addr	port	addr	port	enum	string	string
1422544632.828299	ClCYvR9ovRW3wFZv5	127.0.0.29	15260	127.0.0.21	389	tcp	SPICY_LDAP_TCP	error while parsing LDAP message
```

**Comment by mmguero** [comment]
I ran it through all of the PCAPs I have across various protocols (several thousand), that `krb5.pcap` was the only false positive that triggered from my set.

**Comment by keithjjones** [comment]
> I ran it through all of the PCAPs I have across various protocols (several thousand), that `krb5.pcap` was the only false positive that triggered from my set.

@mmguero I think you found a true positive!  I wired in CLDAP on the UDP side, and that's what Wireshark thinks it is too.  

image

**Comment by mmguero** [comment]
haha I guess I should have looked harder at my traffic. Sorry about the false false positive.

**Comment by keithjjones** [comment]
> haha I guess I should have looked harder at my traffic. Sorry about the false false positive.

I showed you the UDP side.  There is also true LDAP on TCP 389 according to wireshark and your original logs, but it looks like there is a parsing error from the spicy analyzer.  

I read TCP in your logs, but typed UDP in my first screenshot. lol.  Sorry 'bout that.  

image

**Comment by mmguero** [comment]
Cool. Yeah I endorse this PR (not that y'all need my endorsement), seems like it'll do a good job expanding how much of this kind of traffic we'll be able to see.

**Comment by keithjjones** [comment]
Note, I figured out an extra byte and this works better:  https://github.com/zeek/spicy-ldap/pull/4

**Pull request merged**